src/SINDYPI_fit.jl

```diff
@@ -0,0 +1,457 @@
+##FIT FUNCTION
+
+function fit!(X::AbstractArray, Y::AbstractArray, maxiter, lambda, Polyorder, Trigorder)
+    using LinearAlgebra
+    # Compute awesome stuff here
+    s = []
+    for i = 1:1:size(X,2)
+        push!(s, string("x", string(i)))
+    end
+
+number_states_guess = size(X, 2)
+
+Store_Result = Array{Any}(undef, number_states_guess)
+Index_nonzero = Array{Any}(undef, number_states_guess)
+Xi = Array{Any}(undef, number_states_guess)
+
+
+ for w = 1:number_states_guess
+    npoly = Polyorder[w]
+    ntrig = Trigorder[w]
+
+    nvar = size(X, 2)
+
+    basis = ones(size(X,1), 1)
+
+    if npoly >=1
+      basis = hcat(basis, X)
+    end
+
+    if npoly >= 2
+        for i in 1:nvar
+            for j in i:nvar
+                basis = hcat(basis, (X[:, i] .* X[:, j]) )
+            end
+        end
+    end
+
+
+    if npoly >= 3
+        for i in 1:nvar
+            for j in i:nvar
+                for k in j:nvar
+                basis = hcat(basis, (X[:,i] .* X[:, j] .* X[:, k]) )
+            end
+            end
+        end
+    end
+
+
+    if npoly >= 4
+        for i in 1:nvar
+            for j in i:nvar
+                for k in j:nvar
+                    for l in k:nvar
+                basis =  hcat(basis, (X[:,i].* X[:, j] .* X[:,k] .* X[:,l]))
+            end
+            end
+        end
+    end
+end
+
+if npoly >= 5
+    for i in 1:nvar
+        for j in i:nvar
+            for k in j:nvar
+                for l in k:nvar
+                    for m in l:nvar
+            basis =  hcat(basis, (X[:,i].* X[:, j] .* X[:,k] .* X[:,l] .* X[:,m]))
+        end
+        end
+    end
+end
+end
+end
+
+if npoly >= 6
+    for i in 1:nvar
+        for j in i:nvar
+            for k in j:nvar
+                for l in k:nvar
+                    for m in l:nvar
+                        for n in m:nvar
+            basis =  hcat(basis, (X[:,i].* X[:, j] .* X[:,k] .* X[:,l] .* X[:,m] .* X[:,n]))
+        end
+        end
+    end
+end
+end
+end
+end
+
+if ntrig >=1
+    basis = hcat(basis, sin.(X))
+    basis = hcat(basis, cos.(X))
+end
+
+if ntrig >=2
+    basis = hcat(basis, cos.(X) .* cos.(X))
+end
+
+if ntrig >=3
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (X[:, i] .* X[:, i] .* sin.(X[:, j]) .* cos.(X[:, j]) ))
+        end
+    end
+end
+
+if ntrig >=4
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (X[:, i] .* X[:, i] .* sin.(X[:, j])))
+        end
+    end
+
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (X[:, i] .* X[:, i] .* cos.(X[:, j])))
+        end
+    end
+
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (sin.(X[:, i]) .* cos.(X[:, j]) ))
+        end
+    end
+end
+
+    pin = size(basis,2)

```

Comment:
I found this very confusing to read and process, so here are my original comments.

```
    # Measures length of entire basis
    pin = size(basis, 2)
    # Copy to get the derivative free basis
    basis_noderivative = deepcopy(basis[:, 1:pin]) # 
    # Copy for further use
    basis_deep = zeros(size(basis, 1), 2 * size(basis, 2))
    # Fill first part with derivative free
    basis_deep[:, 1:size(basis, 2)] = basis_noderivative
    # Fill second part with basis * choosen derivative
    for k = 1:pin
        basis_deep[:,pin + k] = dx[:, w] .* basis_noderivative[:, k]
    end
    # Same basis with a one variable less
    basis_right = zeros(size(basis_deep, 1), size(basis_deep, 2) - 1)
    # Fill with derivative free basis
    basis_right[:, 1:size(basis_noderivative, 2)] .= basis_deep[:, 1:size(basis_noderivative, 2)]
    # Again ?
    for i = 1:size(basis_noderivative, 2)
        basis_right[:, i] = basis_deep[:,i]
    end
    # Fill second part
    for i = size(basis_noderivative, 2) + 1:size(basis_deep, 2) - 1
        basis_right[:, i] = basis_deep[:,i + 1]
    end
    # Define the first guess
    basis_guess = dx[:, w]
    return (basis_right, basis_guess)
```

Some steps here are doubled and can be kept out.